src/XRP/network-clients/web-socket-network-client.ts

```diff
@@ -0,0 +1,155 @@
+    this.waiting = new Map()
+
+    this.messageCallbacks = new Map()
+    this.messageCallbacks.set('response', (data: WebSocketResponse) => {
+      const dataStatusResponse = data as WebSocketStatusResponse
+      this.waiting.set(dataStatusResponse.id, data)
+    })
+    this.messageCallbacks.set('transaction', this.handleTransaction)
+
+    this.socket.addEventListener('message', (event) => {
+      const parsedData = JSON.parse(event.data) as WebSocketResponse
+      const dataType = parsedData.type
+      const callback = this.messageCallbacks.get(dataType)
+      if (callback) {
+        callback(parsedData)
+      } else {
+        this.handleErrorMessage(
+          'Received unhandlable message: ' + (event.data as string),
+        )
+      }
+    })
+
+    this.socket.addEventListener('close', (event) => {
+      this.handleErrorMessage(
+        'Web socket disconnected, ' + (event.reason as string),
+      )
+    })
+
+    this.socket.addEventListener('error', (event) => {
+      this.handleErrorMessage('Error: ' + (event as string))
+    })
+  }
+
+  /**
+   * Properly handles incoming transactions from the web socket.
+   *
+   * @param data The web socket response received from the web socket.
+   */
+  private handleTransaction = (data: WebSocketResponse) => {
+    const dataTransaction = data as WebSocketTransactionResponse
+    const destinationAccount = dataTransaction.transaction.Destination
+    const callback = this.accountCallbacks.get(destinationAccount)
+    if (callback) {
+      callback(dataTransaction)
+    } else {
+      throw new XrpError(
+        XrpErrorType.Unknown,
+        'Received a transaction for an account that has not been subscribed to: ' +
+          destinationAccount,
+      )
+    }
+  }
+
+  /**
+   * Sends an API request over the web socket connection.
+   *
+   * @param request The object to send over the web socket.
+   * @returns The API response from the web socket.
+   */
+  private async sendApiRequest(
+    request: WebSocketRequestOptions,
+  ): Promise<WebSocketStatusResponse> {
+    while (this.socket.readyState === 0) {
+      await sleep(5)
+    }
+    if (this.socket.readyState !== 1) {
+      throw new XrpError(XrpErrorType.Unknown, 'Socket is closed/closing')
+    }
+    this.socket.send(JSON.stringify(request))
+
+    this.waiting.set(request.id, undefined)
+    while (this.waiting.get(request.id) === undefined) {
+      await sleep(5)
+    }
+    const response = this.waiting.get(request.id)
+    this.waiting.delete(request.id)
+
+    return response as WebSocketStatusResponse
+  }
+
+  /**
+   * Subscribes to incoming transactions to a given account.
+   * @see https://xrpl.org/monitor-incoming-payments-with-websocket.html
+   *
+   * @param id The ID used for the subscription.
+   * @param callback The function called whenever a new transaction is received.
+   * @param account The account from which to subscribe to incoming transactions, encoded as a classic address.
```

Comment:
I was thinking that if you added multiple accounts, you'd probably want a different callback for each account, rather than having them all feed through the same function (and a list of functions is a little messy). 

I certainly can do that, but I could also just add another method (maybe in another PR?) for the option to do multiple accounts.

Reply:
That may be true... I could also see the case where we'd want the same callback for any account, where maybe we update something for the account in some sort of visualizer or something, in an identical way for a variety of accounts.  However, I think it would make perfect sense to do this as a separate API in the future so we can roll with this for now.